File: lib/doit/tasks/progress.ex
defmodule DoIt.Tasks.Progress do
  @moduledoc """
  Pure progress roll-up rules (ProductSpec § Roll-up Progress).

  Operates on `%DoIt.Tasks.Task{}` structs whose `:children` association is
  preloaded recursively (or set to `[]` for leaves). No Repo access — easy to
  test with hand-built structs.

  Rules — **leaf average** (AbstractSpoon-style default):
    * Leaf  → `manual_progress` (clamped to 0..100). Leaf `status: "done"`
      snaps to 100.
    * Branch → plain average over **all descendant leaves** — every leaf
      counts one unit, wherever it sits in the subtree:

          sum(leaf_progress) / leaf_count

      A subtree's pull on its ancestors is its leaf count; decomposing a
      branch further is how it comes to matter more (ProductSpec § Durable
      Principles). (The original single-level average — each direct child
      one unit — is available as the `:single_level` mode, a per-initiative
      setting.)
    * Branch status does NOT snap to 100; progress is always derived from
      leaves so a `done` parent that gains an incomplete child reflects the
      truth. Status reconciliation lives in `DoIt.Tasks`.
    * A branch with no children is treated as a leaf.

  Result is always an integer in 0..100.

  `leaf_value/1` and `average/1` below are the two public building blocks —
  a chain-scoped ancestor recompute (`DoIt.Tasks`) composes them directly
  against rows it fetched itself, instead of assembling a full tree, so
  there's still exactly one implementation of "what a leaf is worth" and
  "how values combine," reused by both the whole-tree walk here and any
  scoped caller.
  """

  alias DoIt.Tasks.Task

  @spec compute(Task.t(), :leaf_average | :single_level) :: integer()
  def compute(task, mode \\ :leaf_average)
  def compute(%Task{} = task, :leaf_average), do: task |> leaf_values() |> average()
  def compute(%Task{} = task, :single_level), do: single_level_value(task)

  @doc """
  Compute every node's rolled-up value in one pass over an assembled tree
  (a list of root nodes). Returns `%{task_id => value}` — the whole-initiative
  reconcile diffs persisted values against this map. One traversal: leaf
  averages aren't derivable per-level from children's persisted percentages
  (they'd also need each child's leaf count), so roll-up always starts
  from the leaves.
  """
  @spec compute_all([Task.t()], :leaf_average | :single_level) :: %{integer() => integer()}
  def compute_all(tree, mode \\ :leaf_average)

  def compute_all(tree, :leaf_average) when is_list(tree) do
    Enum.reduce(tree, %{}, fn node, acc -> elem(evaluate_into(node, acc), 1) end)
  end

  def compute_all(tree, :single_level) when is_list(tree) do
    Enum.reduce(tree, %{}, fn node, acc -> elem(single_level_into(node, acc), 1) end)
  end

  # --- Leaf average -----------------------------------------------------------

  # All descendant leaf values under a node (the node's own value if it's a leaf).
  defp leaf_values(%Task{children: kids}) when is_list(kids) and kids != [] do
    Enum.flat_map(kids, &leaf_values/1)
  end

  defp leaf_values(%Task{} = task), do: [leaf_value(task)]

  # Same traversal, accumulating every node's value into a map.
  # Returns {leaf_values, acc}.
  defp evaluate_into(%Task{children: kids} = task, acc) when is_list(kids) and kids != [] do
    {values, acc} =
      Enum.reduce(kids, {[], acc}, fn child, {values, acc} ->
        {child_values, acc} = evaluate_into(child, acc)
        {values ++ child_values, acc}
      end)

    {values, Map.put(acc, task.id, average(values))}
  end

  defp evaluate_into(%Task{} = task, acc) do
    value = leaf_value(task)
    {[value], Map.put(acc, task.id, value)}
  end

  # --- Single-level average mode (per-initiative setting) ---------------------
  # The original formula: a branch is the plain average of its DIRECT
  # children's rolled-up values — each child one unit regardless of how many
  # leaves it contains.

  defp single_level_value(%Task{children: kids}) when is_list(kids) and kids != [] do
    kids |> Enum.map(&single_level_value/1) |> average()
  end

  defp single_level_value(%Task{} = task), do: leaf_value(task)

  defp single_level_into(%Task{children: kids} = task, acc) when is_list(kids) and kids != [] do
    {values, acc} =
      Enum.reduce(kids, {[], acc}, fn child, {values, acc} ->
        {value, acc} = single_level_into(child, acc)
        {[value | values], acc}
      end)

    value = average(values)
    {value, Map.put(acc, task.id, value)}
  end

  defp single_level_into(%Task{} = task, acc) do
    value = leaf_value(task)
    {value, Map.put(acc, task.id, value)}
  end

  # --- Shared -------------------------------------------------------------

  @doc """
  A single leaf's contribution to a roll-up: `done` snaps to 100, otherwise
  `manual_progress` clamped to 0..100. Accepts anything with `:status` and
  `:manual_progress` keys — a `%Task{}` or a plain map projected straight off
  a query — so a caller scoping a fetch to just those two columns (the
  ancestor-chain recompute in `DoIt.Tasks`) doesn't need a full struct.
  Public so there's exactly one place that maps a task's raw fields to its
  roll-up value; a childless branch is "treated as a leaf" by feeding its
  own row through this same function (see moduledoc).
  """
  @spec leaf_value(%{status: String.t(), manual_progress: integer() | nil}) :: integer()
  def leaf_value(%{status: "done"}), do: 100
  def leaf_value(%{manual_progress: mp}), do: clamp(mp)

  @doc """
  Average a list of roll-up values into the single rounded, clamped result
  (half-up rounding, 0..100). Shared by both modes — `leaf_average` averages
  leaf values, `single_level` averages direct children's values — so a
  chain-scoped recompute can call this directly instead of re-deriving the
  rounding rule.
  """
  @spec average([integer()]) :: integer()
  def average([]), do: 0

  def average(values) do
    values
    |> Enum.sum()
    |> Decimal.new()
    |> Decimal.div(Decimal.new(length(values)))
    |> Decimal.round(0, :half_up)
    |> Decimal.to_integer()
    |> clamp()
  end

  defp clamp(nil), do: 0
  defp clamp(n) when n < 0, do: 0
  defp clamp(n) when n > 100, do: 100
  defp clamp(n) when is_integer(n), do: n
end

File: assets/js/client/tree/progress.ts
// Roll-up progress, client-side (m04.02 item 1.3.2; ProductSpec §8).
//
// A port of `DoIt.Tasks.Progress`, rule for rule, for DISPLAY PREDICTION ONLY.
// The server owns roll-up truth; this exists so a checkbox the user just
// clicked moves its ancestors' bars in the same frame instead of a round trip
// later (UX_GUARDRAILS §6). When the canonical record lands it replaces what
// was predicted — see `delta.ts`.
//
// Rules:
//   * leaf → `manual_progress` clamped 0..100; `done` snaps to 100.
//   * branch (`leaf_average`, the default) → the plain average over EVERY
//     descendant leaf, so a subtree's pull is its leaf count.
//   * branch (`single_level`) → the average of its direct children's values.
//   * a childless branch is treated as a leaf.
//   * a branch's own status never feeds anyone's progress.

import type { ProgressCalc } from "../api/types.ts";
import type { TaskRecord, TreeModel } from "./model.ts";
import { ancestors, subtreeIds } from "./model.ts";

/** One leaf's contribution: `done` is 100, otherwise manual progress, clamped. */
export function leafValue(task: Pick<TaskRecord, "status" | "manual_progress">): number {
  if (task.status === "done") return 100;
  return clamp(task.manual_progress);
}

/**
 * Averages roll-up values into one rounded, clamped result. Half-up rounding on
 * exact integer arithmetic, so it matches `Decimal.round(0, :half_up)` on the
 * server without floating point ever getting a vote.
 */
export function average(values: readonly number[]): number {
  if (values.length === 0) return 0;
  const sum = values.reduce((total, value) => total + value, 0);
  const rounded = Math.floor((2 * sum + values.length) / (2 * values.length));
  return clamp(rounded);
}

function clamp(value: number | null | undefined): number {
  if (typeof value !== "number" || Number.isNaN(value)) return 0;
  if (value < 0) return 0;
  if (value > 100) return 100;
  return Math.trunc(value);
}

/** Every descendant leaf's value under `id` — `id`'s own when it has no children. */
function leafValues(model: TreeModel, id: number): number[] {
  const kids = model.childIds[id] ?? [];
  if (kids.length === 0) {
    const record = model.tasks[id];
    return record === undefined ? [] : [leafValue(record)];
  }
  return kids.flatMap((childId) => leafValues(model, childId));
}

function singleLevelValue(model: TreeModel, id: number): number {
  const kids = model.childIds[id] ?? [];
  if (kids.length === 0) {
    const record = model.tasks[id];
    return record === undefined ? 0 : leafValue(record);
  }
  return average(kids.map((childId) => singleLevelValue(model, childId)));
}

/** `id`'s rolled-up value under the Initiative's progress calculation. */
export function computeProgress(model: TreeModel, id: number): number {
  if (model.progressCalc === "single_level") return singleLevelValue(model, id);
  return average(leafValues(model, id));
}

/**
 * How many units `id`'s roll-up averages over — the chevron badge's
 * denominator, and the Initiative header's when `id` is the system root. A
 * childless task has no units.
 */
export function unitCount(model: TreeModel, id: number, calc?: ProgressCalc): number {
  const kids = model.childIds[id] ?? [];
  if ((calc ?? model.progressCalc) === "single_level") return kids.length;
  return kids.reduce((total, childId) => total + leafCount(model, childId), 0);
}

/** How many of `unitCount`'s units are complete. */
export function doneUnitCount(model: TreeModel, id: number, calc?: ProgressCalc): number {
  const kids = model.childIds[id] ?? [];
  if ((calc ?? model.progressCalc) === "single_level") {
    return kids.filter((childId) => singleLevelValue(model, childId) === 100).length;
  }
  return kids.flatMap((childId) => leafValues(model, childId)).filter((value) => value === 100)
    .length;
}

function leafCount(model: TreeModel, id: number): number {
  const kids = model.childIds[id] ?? [];
  if (kids.length === 0) return 1;
  return kids.reduce((total, childId) => total + leafCount(model, childId), 0);
}

/**
 * Recomputes `progress` for `id` and every ancestor above it — the only records
 * a change under `id` can move. Returns the ids whose value actually changed;
 * untouched records keep their identity.
 */
export function predictLineage(
  model: TreeModel,
  id: number,
): { model: TreeModel; affected: number[] } {
  return recompute(model, [id, ...ancestors(model, id)]);
}

/**
 * The same recompute for `id`'s whole subtree as well as its lineage — what a
 * cascade down (`setDone`) needs, since every descendant's own value moved too.
 */
export function predictSubtreeAndLineage(
  model: TreeModel,
  id: number,
): { model: TreeModel; affected: number[] } {
  // Deepest first, so a branch is recomputed after the children it averages.
  const subtree = subtreeIds(model, id).reverse();
  return recompute(model, [...subtree, ...ancestors(model, id)]);
}

/**
 * Recomputes `progress` up from `parentId` — the same walk as `predictLineage`
 * for a parent that may be the system root (a deletion's old parent, say).
 */
export function predictFrom(
  model: TreeModel,
  parentId: number,
): { model: TreeModel; affected: number[] } {
  if (parentId === model.rootId) return { model, affected: [] };
  return predictLineage(model, parentId);
}

/**
 * The Initiative header's own bar: the system root's roll-up, by the same math
 * end to end (ProductSpec §8). Server truth still replaces it on the next
 * delta; this keeps the top of the screen honest in the meantime.
 */
export function predictHeader(model: TreeModel): TreeModel {
  const progress = computeProgress(model, model.rootId);
  const units = unitCount(model, model.rootId);
  if (model.header.progress === progress && model.header.unit_count === units) return model;
  return { ...model, header: { ...model.header, progress, unit_count: units } };
}

function recompute(model: TreeModel, ids: readonly number[]): {
  model: TreeModel;
  affected: number[];
} {
  const changes: Record<number, TaskRecord> = {};
  const affected: number[] = [];
  let current = model;

  for (const id of ids) {
    const record = current.tasks[id];
    if (record === undefined) continue;
    const progress = computeProgress(current, id);
    if (progress === record.progress) continue;
    const updated = { ...record, progress };
    changes[id] = updated;
    affected.push(id);
    // Each level reads the level below, so the change has to be visible.
    current = { ...current, tasks: { ...current.tasks, [id]: updated } };
  }

  if (affected.length === 0) return { model, affected: [] };
  return { model: { ...model, tasks: { ...model.tasks, ...changes } }, affected };
}
